### PR TITLE
New landing page.

### PR DESCRIPTION
**Pull request by isoos**
Missing pieces tracked in #3246.

Screen Shot 2020-01-21 at 11 48 50

### CODE REVIEW

File: app/lib/frontend/handlers/landing.dart [hunk 1]
```diff
@@ -46,13 +47,34 @@ Future<shelf.Response> indexLandingHandler(shelf.Request request) async {
...
-    final topPackages = await topFeaturedPackages();
+    final topPackages =
+        requestContext.isExperimental ? null : await topFeaturedPackages();
+
+    // TODO: implement top rated
+    final mostPopularPackages = requestContext.isExperimental
+        ? await topFeaturedPackages(order: SearchOrder.popularity)
+        : null;
+
+    // TODO: implement top trending
```
Comment (jonasfj) [inline app/lib/frontend/handlers/landing.dart]: No need to track this is TODOs, file an issue... if I recall correctly we've agreed to not implement that as part of the design.. maybe not even this Q.

File: app/lib/frontend/templates/landing.dart [hunk 2]
```diff
@@ -3,26 +3,60 @@
...
+  List<PackageView> mostPopularPackages, // new only
+  List<PackageView> topFlutterPackages, // new only
+  List<PackageView> topDartPackages, // new only
 }) {
-  final hasTagged = taggedPackages != null && taggedPackages.isNotEmpty;
-  final hasTop = topPackages != null && topPackages.isNotEmpty;
+  final isExperimental = requestContext.isExperimental;
+  final hasTagged =
+      !isExperimental && ffPackages != null && ffPackages.isNotEmpty;
+  final hasTop =
```
Comment (jonasfj) [inline app/lib/frontend/templates/landing.dart]: Maybe we should make a local utility function:

`bool nullOrEmpty(List l) => !isExperimental && l == null && l.isNotEmpty`

File: app/lib/frontend/templates/views/page/landing_experimental.mustache [hunk 3]
```diff
@@ -0,0 +1,60 @@
+{{! Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
+    for details. All rights reserved. Use of this source code is governed by a
+    BSD-style license that can be found in the LICENSE file. }}
+
+{{#has_ff}}
```
Comment (jonasfj) [inline app/lib/frontend/templates/views/page/landing_experimental.mustache]: Can't we just make a condition on `ff_mini_list_html`? and avoid the `has_...` variables?

File: app/lib/frontend/templates/landing.dart [hunk 4]
```diff
@@ -3,26 +3,60 @@
...
     'has_tagged': hasTagged,
-    'tagged_minilist_html': hasTagged ? renderMiniList(taggedPackages) : null,
+    'tagged_minilist_html': hasTagged ? renderMiniList(ffPackages) : null,
     'tagged_more_url': '/flutter/favorites',
     'has_top': hasTop,
     'top_minilist_html': hasTop ? renderMiniList(topPackages) : null,
     'top_more_url': urls.searchUrl(),
+    // new design's variables
+    'has_ff': hasFF,
+    'ff_mini_list_html': hasFF ? renderMiniList(ffPackages) : null,
```
Comment (jonasfj) [inline app/lib/frontend/templates/landing.dart]: Maybe make a local utility function:

 `final renderMiniListOrNull = (x) => (x == null || x.isEmpty) ? null : renderMiniList(x);`

**Comment by jonasfj** [comment]
I was a little worried when I saw all the `has...` variables, that seemed a bit messy -- but the logic is simple to follow. Cleaning up with a helper function or two might be nice, but it's a nit :)

File: app/lib/frontend/templates/views/page/landing_experimental.mustache (hunk 3 above)
Comment (isoos) [reply app/lib/frontend/templates/views/page/landing_experimental.mustache]: We could, but in the future non-nullability may break such templates unexpectedly, I think it is better to be explicit if the overhead is small.

**Pull request merged**